src/vmm_vmmd.ml

```diff
@@ -370,17 +390,37 @@ let handle_unikernel_cmd t id = function
          | () -> ());
         Ok (t, `Wait_and_create (id, (id, unikernel_config)))
     end
-  | `Unikernel_restart ->
+  | `Unikernel_restart args ->
     begin
       match Vmm_resources.find_unikernel t.resources id with
-      | None -> stop_create t id
+      | None ->
+        (* TODO: unsure about the semantics here -- a restart on a non-existing
+           unikernel should do what? and on a currently restart(ing-on-failure)
+           one? *)
+        stop_create t id
       | Some unikernel ->
         let* resources = Vmm_resources.remove_unikernel t.resources id in
         let* () = Vmm_resources.check_unikernel resources id unikernel.Unikernel.config in
         (match Vmm_unix.destroy unikernel with
          | exception Unix.Unix_error _ -> ()
          | () -> ());
-        Ok (t, `Wait_and_create (id, (id, unikernel.Unikernel.config)))
+        let config =
+          match args with
+          | None -> unikernel.Unikernel.config
+          | Some (a : Unikernel.arguments) ->
+            { unikernel.Unikernel.config with
+              fail_behaviour = a.fail_behaviour ;
+              cpuid = a.cpuid ;
+              block_devices = a.block_devices ;
+              bridges = a.bridges ;
+              argv = a.argv }
+        in
+        match Vmm_unix.manifest_devices_match ~bridges:config.bridges
+                ~block_devices:config.block_devices
+                config.image
```

Comment:
I think it is preferable that we check the manifest matches before we kill off the existing instance. If they don't match we kill off the existing instance and then don't start a new one.

Reply:
I will write and push a fix.